File: python/lightning/client/_retry.py
from __future__ import annotations

import random

from lightning.client._types import RetryConfig


def compute_backoff(
    attempt: int,
    config: RetryConfig,
) -> float:
    delay = min(config.base_delay * (2 ** attempt), config.max_delay)
    jitter = random.uniform(-config.jitter_factor * delay, config.jitter_factor * delay)
    return max(0.0, delay + jitter)


def should_retry(
    status_code: int,
    attempt: int,
    config: RetryConfig,
) -> bool:
    if attempt >= config.max_retries:
        return False
    return status_code in config.retryable_statuses


def is_connection_error(exception: Exception) -> bool:
    import httpx
    if isinstance(exception, (httpx.ConnectError, httpx.ConnectTimeout)):
        return True
    if isinstance(exception, httpx.ReadTimeout) and "connection" in str(exception).lower():
        return True
    if isinstance(exception, httpx.RemoteProtocolError):
        return True
    return False

File: packages/lightning-client/src/retry.ts
export interface RetryConfig {
  maxRetries: number
  baseDelay: number
  maxDelay: number
  jitterFactor: number
  retryableStatuses: ReadonlySet<number>
}

export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  baseDelay: 0.1,
  maxDelay: 10.0,
  jitterFactor: 0.1,
  retryableStatuses: new Set([429, 502, 503, 504]),
}

export function computeBackoff(attempt: number, config: RetryConfig): number {
  const delay = Math.min(config.baseDelay * Math.pow(2, attempt), config.maxDelay)
  const jitter = (Math.random() * 2 - 1) * config.jitterFactor * delay
  return Math.max(0, delay + jitter)
}

export function shouldRetry(statusCode: number, attempt: number, config: RetryConfig): boolean {
  if (attempt >= config.maxRetries) return false
  return config.retryableStatuses.has(statusCode)
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms * 1000))
}
